### PR TITLE
Feat: Support Granite 3.0 MoE in SGLang

### PR DESCRIPTION
**Pull request by zminglei**
## Motivation
Model Support Issue: https://github.com/sgl-project/sglang/issues/7429
This PR will support GraniteMoeForCausalLM (Granite 3.0 MoE)

## Modifications
The original model implementation comes from vllm: https://github.com/vllm-project/vllm/blob/e2de455c349df8385b18fe447beb6325dcb6af9c/vllm/model_executor/models/granitemoe.py

This PR ports the vllm model into sglang by following this sglang guide: https://docs.sglang.ai/supported_models/support_new_models.html#port-a-model-from-vllm-to-sglang

Below integration tests are performed under 2xH100 GPU (tp=2) & [ibm-granite/granite-3.0-3b-a800m-instruct](https://huggingface.co/ibm-granite/granite-3.0-3b-a800m-instruct)
`python3 -m sglang.launch_server --model-path /shared/public/elr-models/ibm-granite/granite-3.0-3b-a800m-instruct/50ada2f5b1c5b6132121976067803a00b83b0d53/ --trust-remote-code --tp 2`

## Bechmark
The metrics result align with the result on the model's report.
image
**GSM8K**
```
python benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1319 --parallel 1319 --data-path /shared/public
/data/gsm8k/test.jsonl
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [00:16<00:00, 79.68it/s]
Accuracy: 0.519
Invalid: 0.001
Latency: 16.629 s
Output throughput: 12629.558 token/s
```

**MMLU**
```
TIKTOKEN_CACHE_DIR=/shared/user/zminglei/tiktoken_cache_download/ python benchmark/mmlu/bench_sglang.py --data_dir /shared/user/zminglei/data
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 14042/14042 [00:45<00:00, 311.24it/s]
subject: abstract_algebra, #q:100, acc: 0.400
subject: anatomy, #q:135, acc: 0.511
subject: astronomy, #q:152, acc: 0.553
subject: business_ethics, #q:100, acc: 0.470
subject: clinical_knowledge, #q:265, acc: 0.577
subject: college_biology, #q:144, acc: 0.562
subject: college_chemistry, #q:100, acc: 0.360
subject: college_computer_science, #q:100, acc: 0.440
subject: college_mathematics, #q:100, acc: 0.360
subject: college_medicine, #q:173, acc: 0.532
subject: college_physics, #q:102, acc: 0.343
subject: computer_security, #q:100, acc: 0.580
subject: conceptual_physics, #q:235, acc: 0.460
subject: econometrics, #q:114, acc: 0.298
subject: electrical_engineering, #q:145, acc: 0.531
subject: elementary_mathematics, #q:378, acc: 0.339
subject: formal_logic, #q:126, acc: 0.349
subject: global_facts, #q:100, acc: 0.290
subject: high_school_biology, #q:310, acc: 0.629
subject: high_school_chemistry, #q:203, acc: 0.389
subject: high_school_computer_science, #q:100, acc: 0.470
subject: high_school_european_history, #q:165, acc: 0.594
subject: high_school_geography, #q:198, acc: 0.616
subject: high_school_government_and_politics, #q:193, acc: 0.746
subject: high_school_macroeconomics, #q:390, acc: 0.459
subject: high_school_mathematics, #q:270, acc: 0.289
subject: high_school_microeconomics, #q:238, acc: 0.504
subject: high_school_physics, #q:151, acc: 0.298
subject: high_school_psychology, #q:545, acc: 0.721
subject: high_school_statistics, #q:216, acc: 0.431
subject: high_school_us_history, #q:204, acc: 0.603
subject: high_school_world_history, #q:237, acc: 0.624
subject: human_aging, #q:223, acc: 0.610
subject: human_sexuality, #q:131, acc: 0.550
subject: international_law, #q:121, acc: 0.669
subject: jurisprudence, #q:108, acc: 0.611
subject: logical_fallacies, #q:163, acc: 0.632
subject: machine_learning, #q:112, acc: 0.375
subject: management, #q:103, acc: 0.709
subject: marketing, #q:234, acc: 0.744
subject: medical_genetics, #q:100, acc: 0.540
subject: miscellaneous, #q:783, acc: 0.682
subject: moral_disputes, #q:346, acc: 0.566
subject: moral_scenarios, #q:895, acc: 0.239
subject: nutrition, #q:306, acc: 0.601
subject: philosophy, #q:311, acc: 0.585
subject: prehistory, #q:324, acc: 0.580
subject: professional_accounting, #q:282, acc: 0.344
subject: professional_law, #q:1534, acc: 0.379
subject: professional_medicine, #q:272, acc: 0.426
subject: professional_psychology, #q:612, acc: 0.500
subject: public_relations, #q:110, acc: 0.518
subject: security_studies, #q:245, acc: 0.616
subject: sociology, #q:201, acc: 0.682
subject: us_foreign_policy, #q:100, acc: 0.740
subject: virology, #q:166, acc: 0.458
subject: world_religions, #q:171, acc: 0.702
Total latency: 45.204
Average accuracy: 0.504
```
## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [x] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/models/granitemoe.py [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+
+        self.residual_multiplier = config.residual_multiplier
+
+    def forward(
+        self,
+        positions: torch.Tensor,
+        hidden_states: torch.Tensor,
+        forward_batch: ForwardBatch,
+    ) -> torch.Tensor:
+        # Self Attention
```
Comment (byjiang1996) [inline python/sglang/srt/models/granitemoe.py]: nit: move this comment right above self.self_attn

File: python/sglang/srt/models/granitemoe.py [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+        self,
+        config: GraniteConfig,
+        quant_config: Optional[QuantizationConfig] = None,
+        prefix: str = "",
+    ):
+        super().__init__()
+
+        self.config = config
+        self.vocab_size = config.vocab_size
+        self.org_vocab_size = config.vocab_size
```
Comment (byjiang1996) [inline python/sglang/srt/models/granitemoe.py]: nit: remove

File: python/sglang/srt/models/granitemoe.py [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+    def __init__(
+        self,
+        config: GraniteConfig,
+        layer_id: int = 0,
+        quant_config: Optional[QuantizationConfig] = None,
+        prefix: str = "",
+    ) -> None:
+        super().__init__()
+        self.hidden_size = config.hidden_size
+        # Requires transformers > 4.32.0
```
Comment (byjiang1996) [inline python/sglang/srt/models/granitemoe.py]: nit: remove this comment. transformer v4.32 is very old. https://github.com/sgl-project/sglang/pull/7907#discussion_r2196442887

File: python/sglang/srt/models/granitemoe.py [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+            "v_proj",
+        ],
+    }
+
+    # LoRA specific attributes
+    embedding_modules = {
+        "embed_tokens": "input_embeddings",
+        "lm_head": "output_embeddings",
+    }
+    embedding_padding_modules = ["lm_head"]
```
Comment (byjiang1996) [inline python/sglang/srt/models/granitemoe.py]: nit: Remove. Seems not required in sglang.

File: python/sglang/srt/models/granitemoe.py [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+
+    def __init__(
+        self,
+        config: GraniteConfig,
+        quant_config: Optional[QuantizationConfig] = None,
```
Comment (byjiang1996) [inline python/sglang/srt/models/granitemoe.py:295]: Can you test the model with `--quantization fp8`?

Ref: https://github.com/sgl-project/sglang/pull/7907#discussion_r2196445830

File: python/sglang/srt/models/granitemoe.py [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+            quant_config=quant_config,
+            tp_size=tp_size,
+            prefix=f"{prefix}.experts",
+        )
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        # NOTE: hidden_states can have either 1D or 2D shape.
+        orig_shape = hidden_states.shape
+        hidden_states = hidden_states.view(-1, self.hidden_size)
+        # router_logits: (num_tokens, n_experts)
```
Comment (hebiao064) [inline python/sglang/srt/models/granitemoe.py]: remove this line

File: python/sglang/srt/models/granitemoe.py (hunk 1 above)
Comment (zminglei) [reply python/sglang/srt/models/granitemoe.py]: Sounds good.

File: python/sglang/srt/models/granitemoe.py (hunk 5 above)
Comment (zminglei) [reply python/sglang/srt/models/granitemoe.py:295]: Sure, re-tested it with --quntization fp8 and got this:
```
(venv) jobuser [ /shared/user/zminglei/repos/sglang ]$ python benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1319 --parallel 1319 --data-path /shared/public/data/gsm8k/test.jsonl 
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [00:31<00:00, 41.62it/s]
Accuracy: 0.497
Invalid: 0.001
Latency: 32.834 s
Output throughput: 6371.998 token/s
(venv) jobuser [ /shared/user/zminglei/repos/sglang ]$ TIKTOKEN_CACHE_DIR=/shared/user/zminglei/tiktoken_cache_download/ python benchmark/mmlu/bench_sglang.py --data_dir /shared/user/zminglei/data
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 14042/14042 [00:39<00:00, 353.78it/s]
subject: abstract_algebra, #q:100, acc: 0.400
subject: anatomy, #q:135, acc: 0.496
subject: astronomy, #q:152, acc: 0.539
subject: business_ethics, #q:100, acc: 0.500
subject: clinical_knowledge, #q:265, acc: 0.615
subject: college_biology, #q:144, acc: 0.604
subject: college_chemistry, #q:100, acc: 0.390
subject: college_computer_science, #q:100, acc: 0.460
subject: college_mathematics, #q:100, acc: 0.330
subject: college_medicine, #q:173, acc: 0.543
subject: college_physics, #q:102, acc: 0.343
subject: computer_security, #q:100, acc: 0.580
subject: conceptual_physics, #q:235, acc: 0.455
subject: econometrics, #q:114, acc: 0.316
subject: electrical_engineering, #q:145, acc: 0.510
subject: elementary_mathematics, #q:378, acc: 0.362
subject: formal_logic, #q:126, acc: 0.381
subject: global_facts, #q:100, acc: 0.280
subject: high_school_biology, #q:310, acc: 0.635
subject: high_school_chemistry, #q:203, acc: 0.369
subject: high_school_computer_science, #q:100, acc: 0.480
subject: high_school_european_history, #q:165, acc: 0.600
subject: high_school_geography, #q:198, acc: 0.616
subject: high_school_government_and_politics, #q:193, acc: 0.731
subject: high_school_macroeconomics, #q:390, acc: 0.444
subject: high_school_mathematics, #q:270, acc: 0.296
subject: high_school_microeconomics, #q:238, acc: 0.508
subject: high_school_physics, #q:151, acc: 0.278
subject: high_school_psychology, #q:545, acc: 0.732
subject: high_school_statistics, #q:216, acc: 0.398
subject: high_school_us_history, #q:204, acc: 0.603
subject: high_school_world_history, #q:237, acc: 0.641
subject: human_aging, #q:223, acc: 0.601
subject: human_sexuality, #q:131, acc: 0.534
subject: international_law, #q:121, acc: 0.645
subject: jurisprudence, #q:108, acc: 0.593
subject: logical_fallacies, #q:163, acc: 0.632
subject: machine_learning, #q:112, acc: 0.312
subject: management, #q:103, acc: 0.699
subject: marketing, #q:234, acc: 0.731
subject: medical_genetics, #q:100, acc: 0.550
subject: miscellaneous, #q:783, acc: 0.674
subject: moral_disputes, #q:346, acc: 0.552
subject: moral_scenarios, #q:895, acc: 0.254
subject: nutrition, #q:306, acc: 0.582
subject: philosophy, #q:311, acc: 0.566
subject: prehistory, #q:324, acc: 0.583
subject: professional_accounting, #q:282, acc: 0.372
subject: professional_law, #q:1534, acc: 0.377
subject: professional_medicine, #q:272, acc: 0.445
subject: professional_psychology, #q:612, acc: 0.480
subject: public_relations, #q:110, acc: 0.527
subject: security_studies, #q:245, acc: 0.608
subject: sociology, #q:201, acc: 0.667
subject: us_foreign_policy, #q:100, acc: 0.750
subject: virology, #q:166, acc: 0.428
subject: world_religions, #q:171, acc: 0.673
Total latency: 39.768
Average accuracy: 0.502
```

File: python/sglang/srt/models/granitemoe.py (hunk 6 above)
Comment (zminglei) [reply python/sglang/srt/models/granitemoe.py]: Sounds good.

File: python/sglang/srt/models/granitemoe.py (hunk 4 above)
Comment (zminglei) [reply python/sglang/srt/models/granitemoe.py]: Good point, removed the useless variables here.

File: python/sglang/srt/models/granitemoe.py (hunk 3 above)
Comment (zminglei) [reply python/sglang/srt/models/granitemoe.py]: Sounds good

File: python/sglang/srt/models/granitemoe.py (hunk 2 above)
Comment (zminglei) [reply python/sglang/srt/models/granitemoe.py]: sure, removed it

**Review by byjiang1996 — APPROVED** [review]
LGTM

File: python/sglang/srt/models/granitemoe.py [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+from sglang.srt.layers.radix_attention import RadixAttention
+from sglang.srt.layers.rotary_embedding import get_rope
+from sglang.srt.layers.vocab_parallel_embedding import (
+    ParallelLMHead,
+    VocabParallelEmbedding,
+)
+from sglang.srt.model_executor.forward_batch_info import ForwardBatch
+from sglang.srt.utils import add_prefix
+
+from . import mixtral
```
Comment (hebiao064) [inline python/sglang/srt/models/granitemoe.py]: `from sglang.path.to.module import mixtral`

**Comment by hebiao064** [comment]
LGTM with one minor comment

File: python/sglang/srt/models/granitemoe.py (hunk 7 above)
Comment (zminglei) [reply python/sglang/srt/models/granitemoe.py]: Sounds good, updated it, ptal

**Review by hebiao064 — APPROVED** [review]
